src/Stratis.Bitcoin/Consensus/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,105 @@
+        {
+            this.chain = chain;
+            this.initialBlockDownloadState = initialBlockDownloadState;
+            this.consensusManager = consensusManager;
+            this.peerBanning = peerBanning;
+            this.loggerFactory = loggerFactory;
+            this.consensusFactory = new PosConsensusFactory();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        /// <inheritdoc />
+        /// <summary>
+        /// Processes and incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        protected override async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            await base.OnMessageReceivedAsync(peer, message).ConfigureAwait(false);
+            switch (message.Message.Payload)
+            {
+                case ProvenHeadersPayload provenHeaders:
+                    await this.ProcessHeadersAsync(peer, provenHeaders.Headers.Cast<BlockHeader>().ToList()).ConfigureAwait(false);
+                    break;
+
+                case GetProvenHeadersPayload getHeaders:
+                    await this.ProcessGetHeadersAsync(peer, getHeaders).ConfigureAwait(false);
+                    break;
+            }
+        }
+
+        /// <inheritdoc />
+        /// <summary>Constructs the proven headers payload from locator to consensus tip.</summary>
+        /// <param name="locator">Block locator.</param>
+        /// <param name="hashStop">Hash of the block after which constructing headers payload should stop.</param>
+        /// <param name="lastHeader"><see cref="T:NBitcoin.ProvenBlockHeader" /> of the last header that was added to the <see cref="T:Stratis.Bitcoin.P2P.Protocol.Payloads.ProvenHeadersPayload" />.</param>
+        /// <returns><see cref="T:Stratis.Bitcoin.P2P.Protocol.Payloads.ProvenHeadersPayload" /> with headers from locator towards consensus tip or <c>null</c> in case locator was invalid.</returns>
+        protected override Payload ConstructHeadersPayload(BlockLocator locator, uint256 hashStop, out ChainedHeader lastHeader)
+        {
+            ChainedHeader fork = this.chain.FindFork(locator);
+
+            lastHeader = null;
+
+            if (fork == null)
+            {
+                this.logger.LogTrace("(-)[INVALID_LOCATOR]:null");
+                return null;
+            }
+
+            var headers = new ProvenHeadersPayload();
+            foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+            {
+                var posBock = new PosBlock(header.Header);
```

Comment:
this is probably wrong, the factory should have a way to create a header without a block